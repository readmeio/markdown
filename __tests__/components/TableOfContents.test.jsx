const React = require('react');
const { shallow, mount } = require('enzyme');

const markdown = require('../../index');
const TableOfContents = require('../../components/TableOfContents');

const { GlossaryContext, VariablesContext } = markdown.utils;
const variables = {
  defaults: [{ test: 'Default Value' }],
  user: { test: 'User Override' },
};
const glossaryTerms = [
  {
    term: 'demo',
    definition: 'a thing that breaks on presentation',
  },
];

describe('Table of Contents', () => {
  it('should have a header', () => {
    const toc = shallow(
      <TableOfContents>
        <li>Heading 1</li>
      </TableOfContents>
    );

    expect(toc.find('li').first().text()).toBe('Table of Contents');
  });

  it('generates TOC from headings', () => {
    const txt = '# Heading Zed\n\n# Heading One';
    const ast = markdown.reactProcessor().parse(txt);
    const toc = markdown.reactTOC(ast);
    const dom = mount(toc);

    const items = dom.find('li > a[href]').not('[href=""]');
    expect(items).toHaveLength(2);
  });

  it('includes two heading levels', () => {
    const txt = '# Heading Zed\n\n## Subheading One\n\n### Deep Heading Two';
    const ast = markdown.reactProcessor().parse(txt);
    const toc = markdown.reactTOC(ast);
    const dom = mount(toc);

    const items = dom.find('li > a[href]').not('[href=""]');
    expect(items).toHaveLength(2);
    expect(dom.html()).toMatchSnapshot();
  });

  it('normalizes root depth level', () => {
    const txt = '##### Heading Zed\n\n###### Subheading Zed';
    const ast = markdown.reactProcessor().parse(txt);
    const toc = markdown.reactTOC(ast);
    const dom = mount(toc);

    const items = dom.find('li > a[href]').not('[href=""]');
    expect(items).toHaveLength(2);
  });

  it('includes variables', () => {
    const txt = '# Heading <<test>>';
    const ast = markdown.reactProcessor().parse(txt);
    const toc = markdown.reactTOC(ast);
    const dom = mount(<VariablesContext.Provider value={variables}>{toc}</VariablesContext.Provider>);

    const items = dom.find('li > a[href]').not('[href=""]');
    expect(items.text()).toStrictEqual(`Heading ${variables.user.test}`);
  });

  it('includes glossary items', () => {
    const txt = '# Heading <<glossary:demo>>';
    const ast = markdown.reactProcessor().parse(txt);
    const toc = markdown.reactTOC(ast);
    const dom = mount(<GlossaryContext.Provider value={glossaryTerms}>{toc}</GlossaryContext.Provider>);

    const items = dom.find('li > a[href]').not('[href=""]');
    expect(items.text()).toMatch(`Heading ${glossaryTerms[0].term}`);
  });
});
