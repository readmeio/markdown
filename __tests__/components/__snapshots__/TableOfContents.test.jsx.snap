// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table of Contents includes two heading levels 1`] = `
"<nav><ul class=\\"toc-list\\"><li><a class=\\"tocHeader\\"><i class=\\"icon icon-text-align-left\\"></i>Table of Contents</a></li><li class=\\"toc-children\\"><ul>
<li>
<a href=\\"#heading-zed\\">Heading Zed</a>
<ul>
<li><a href=\\"#subheading-one\\">Subheading One</a></li>
</ul>
</li>
</ul></li></ul></nav>"
`;
